### PR TITLE
Implement API for setting RM

### PR DESCRIPTION
**Pull request by lukebarnard1**
This is now stored on the server with similar treatment to RRs. The server will only store the specified eventId as the current read marker for a room if the event is ahead in the stream when compared to the existing RM. The exception is when the RM has never been set for this room for this user, in which case the event ID will be stored as the RM without any comparison.

This API also allows for an optional RR event ID to be sent in the same request. This is because it might be the common case for some clients to update the RM at the same time as updating the RR.

See design: https://docs.google.com/document/d/1UWqdS-e1sdwkLDUY0wA4gZyIkRp-ekjsLZ8k6g_Zvso/edit

See server-side PR: https://github.com/matrix-org/synapse/pull/2120

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -843,6 +843,43 @@ MatrixClient.prototype.setRoomAccountData = function(roomId, eventType,
...
+ * @param {string} eventId ID of the event that has been read
+ * @param {string} rrEvent the event tracked by the read receipt. This is here for
+ * convenience because the RR and the RM are commonly updated at the same time as each
+ * other. Optional.
+ * @param {module:client.callback} callback Optional.
+ * @return {module:client.Promise} Resolves: TODO
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.setRoomReadMarker = function(roomId, eventId, rrEvent, callback) {
+    if (typeof rrEventId === 'function') {
```
Comment (richvdh) [inline src/client.js]: why is this a thing? some functions have this sort of hackery, but it's just for backwards compat. You shouldn't need it for a new func

File: src/client.js [hunk 2]
```diff
@@ -843,6 +843,43 @@ MatrixClient.prototype.setRoomAccountData = function(roomId, eventType,
 };
 
 /**
+ * @param {string} roomId ID of the room that has been read
+ * @param {string} eventId ID of the event that has been read
+ * @param {string} rrEvent the event tracked by the read receipt. This is here for
+ * convenience because the RR and the RM are commonly updated at the same time as each
+ * other. Optional.
+ * @param {module:client.callback} callback Optional.
```
Comment (richvdh) [inline src/client.js]: we're deprecating `callback`s and not adding them for new funcs

File: src/client.js [hunk 3]
```diff
@@ -843,6 +843,43 @@ MatrixClient.prototype.setRoomAccountData = function(roomId, eventType,
...
+ * @return {module:client.Promise} Resolves: TODO
```
Comment (richvdh) [inline src/client.js]: can you find something better than this please

File: src/client.js [hunk 4]
```diff
@@ -843,6 +843,43 @@ MatrixClient.prototype.setRoomAccountData = function(roomId, eventType,
 };
 
 /**
```
Comment (richvdh) [inline src/client.js]: what does it DOOOOOOOOOO?

File: src/client.js (hunk 3 above)
Comment (lukebarnard1) [inline src/client.js]: Would it to be terrible to do `(See @http.authedRequest)` or whatever?

File: src/client.js (hunk 1 above)
Comment (lukebarnard1) [inline src/client.js]: `rrEvent` and `callback` are both optional.

File: src/client.js (hunk 2 above)
Comment (lukebarnard1) [inline src/client.js]: Oh. OK

**Comment by lukebarnard1** [comment]
>can you move as much of this as possible into a low-level func in base-apis.js

I'm guessing you mean 
```js
const room = this.getRoom(roomId);
if (room) {
    room._addLocalEchoReceipt(this.credentials.userId, rrEvent, "m.read");
}
```
but `getRoom` is in `client.js` so it'd feel wrong to move it.

**Comment by richvdh** [comment]
No, I mean the all the rest of it.

File: src/base-apis.js [hunk 5]
```diff
@@ -470,6 +470,33 @@ MatrixBaseApis.prototype.roomInitialSync = function(roomId, limit, callback) {
...
+ * Set a marker to represent which event the user was last reading in a room. This can
+ * be retrieved from room account data and displayed as a horizontal line in the client
+ * that is visually distinct to the position of the user's own read receipt.
+ * @param {string} roomId ID of the room that has been read
+ * @param {string} rmEventId ID of the event that has been read
+ * @param {string} rrEventId ID of the event tracked by the read receipt. This is here
+ * for convenience because the RR and the RM are commonly updated at the same time as
+ * each other. Optional.
+ * @return {module:client.Promise} Resolves: (@see module:http-api.authedRequest)
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
```
Comment (richvdh) [inline src/base-apis.js]: it never returns a MatrixError; it always returns a Promise.

Sorry, I should have been clearer about what to put here. There are a lot of historical functions (including authedRequest) which are basically full of shit. Note also that `httpApi.opts.onlyData` is true as far as base-apis is concerned, so authedRequest always just returns the result of the api call, so linking to that isn't particularly helpful.

Basically the best thing to do here is just have one `@return` line, along the lines of:

```
@return {module:client.Promise} resolves to the result object for the API call.
```

If you have the time and energy to document the shame of that object, so much the better, but generally we expect people to go and read the spec.

File: src/base-apis.js [hunk 6]
```diff
@@ -470,6 +470,33 @@ MatrixBaseApis.prototype.roomInitialSync = function(roomId, limit, callback) {
     );
 };
 
+/**
+ * Set a marker to represent which event the user was last reading in a room. This can
```
Comment (richvdh) [inline src/base-apis.js]: it's not the event the user was last reading though, is it?

File: src/client.js [hunk 7]
```diff
@@ -1201,6 +1201,33 @@ MatrixClient.prototype.sendReadReceipt = function(event, callback) {
...
+ * that is visually distinct to the position of the user's own read receipt.
+ * @param {string} roomId ID of the room that has been read
+ * @param {string} eventId ID of the event that has been read
+ * @param {string} rrEvent the event tracked by the read receipt. This is here for
+ * convenience because the RR and the RM are commonly updated at the same time as each
+ * other. Optional.
+ * @return {module:client.Promise} Resolves: (@see module:http-api.authedRequest)
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.setRoomReadMarker = function(roomId, eventId, rrEvent) {
```
Comment (richvdh) [inline src/client.js]: let's not distinguish these two by the presence or absence of 's'.

How about calling the other `setRoomReadMarkersHttpRequest` or something (we have a `_sendEventHttpRequest` for distinction from `_sendEvent`)

File: src/client.js [hunk 8]
```diff
@@ -1201,6 +1201,33 @@ MatrixClient.prototype.sendReadReceipt = function(event, callback) {
     return this.sendReceipt(event, "m.read", callback);
 };
 
+/**
```
Comment (richvdh) [inline src/client.js]: jsdocs need tweaks as above

File: src/interactive-auth.js [hunk 9]
```diff
@@ -318,6 +318,13 @@ InteractiveAuth.prototype = {
         }
         this._currentStage = nextStage;
 
+        if (nextStage == 'm.login.dummy') {
```
Comment (richvdh) [inline src/interactive-auth.js]: ?

**Comment by lukebarnard1** [comment]
Now that I've changed the base of the merge, those changes to `release.sh` and `interactive-auth.js` should've disappeared
